src/AWS.Deploy.Orchestration/Orchestrator.cs

```diff
@@ -163,7 +165,19 @@ public async Task DeployRecommendation(CloudApplication cloudApplication, Recomm
 
                     var projFiles = _directoryManager.GetProjFiles(cdkProject);
                     var cdkVersion = _cdkVersionDetector.Detect(projFiles);
-                    await _cdkManager.EnsureCompatibleCDKExists(Constants.CDK.DeployToolWorkspaceDirectoryRoot, cdkVersion);
+
+                    try
+                    {
+                        await s_cdkManagerSemaphoreSlim.WaitAsync();
+                        await _cdkManager.EnsureCompatibleCDKExists(Constants.CDK.DeployToolWorkspaceDirectoryRoot, cdkVersion);
+                    }
+                    finally
+                    {
+                        if (s_cdkManagerSemaphoreSlim.CurrentCount == 0)
```

Comment:
Why would we not always call `Release`? Seems like the `if` statement is just adding complexity but if we enter the block we should always release on the way out.

Reply:
Because release can throw an exception (SemaphoreFullException)

https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphore.release?view=net-5.0

Reply:
This can happen if my WaitAsync call fails and in finally block, it tries to release the semaphore.

Reply:
The exception would be because you have a design issue where you are calling release more often then entering the lock. If you did have that design issue this `if` statement would be hiding that issue.